framework/include/mesh/MooseMesh.h

```diff
@@ -71,7 +71,35 @@ class QpMap
 class FaceInfo
 {
 public:
-  FaceInfo(const Elem * elem, const Elem * neighbor);
+  FaceInfo(const Elem * elem, unsigned int side, const Elem * neighbor);
+
+  /**
+   * type of this face for each variables
+   * each variable potentially has its own domain and boundary
+   * INTERNAL: internal to the domain of the variable
+   * INVALID: variable is not defined on either subdomains (left/right)
+   * LEFT_DIRICHLET: face defines a Dirichlet BC for this variable and variable
+   *                 is defined on the left element
+   * RIGHT_DIRICHLET: face defines a Dirichlet BC for this variable and variable
+   *                  is defined on the right element
+   * LEFT_INTEGRATED: face defines an Integrated BC for this variable and variable
+   *                 is defined on the left element
+   * RIGHT_INTEGRATED: face defines an Integrated BC for this variable and variable
+   *                  is defined on the right element
+   * NATURAL_BOUNDARY: face is a boundary for this variable but no boundary conditions
+   *                   is defined
```

Comment:
I think all we need this enum to have is INTERNAL, INVALID, LEFT, RIGHT.  And then a separate boolean (maybe) that says whether or not there are any dirichlet BCs on the boundary.  That way we split out these orthogonal concerns - whether or not neighbor element/values exist for a variable at a face vs whether or not we have any objects that need values on *both* sides of the face -- the first case is necessary for correctness and the second is just for performance (e.g. to skip reiniting neighbor ghost values for a variable if there is only a flux BC).

Reply:
I think you're right. `NATURAL` occurs naturally if you don't define a BC.
`INVALID` is a bad name I think, maybe `DO_NOTHING`?

Reply:
How about `enum class VarFaceNeighbors {BOTH, LEFT, RIGHT, NONE}` specifying on which side(s) the solution is defined.

Reply:
Or maybe NEITHER instead of NONE